STYLE.md

```diff
@@ -0,0 +1,36 @@
+# C++ coding style
+
+* The Envoy source code is formated using clang-format. Thus all white space, etc.
+  issues are taken care of automatically. The Travis tests will automatically check
+  the code format and fail. There are make targets that can both check the format
+  (check_format) as well as fix the code format for you (fix_format).
+* Beyond code formatting, for the most part Envoy uses the
+  [Google C++ style guidelines](https://google.github.io/styleguide/cppguide.html).
+  The following section covers the major areas where we deviate from the Google
+  guidelines.
+
+# Deviations from Google C++ style guidelines
+
+* Exceptions are allowed and encouraged where appropriate.
+* References are always preferred over pointers when the reference cannot be null. This
+  includes both const and non-const references.
+* Function names using camel case starting with a lower case letter (e.g., "doFoo()").
+* Struct/Class member variables have a '\_' postfix (e.g., "int foo\_;").
+* 100 columns is the line limit.
+* OOM events (both memory and FDs) are considered fatal crashing errors.
```

Comment:
Can we add some doco on the use of assertions? E.g.

All error code returns should be checked. It's acceptable to turn failures into crash semantics via `RELEASE_ASSERT(condition)` if there is no other sensible behavior, e.g. in OOM (memory/FD) scenarios. Crash semantics are preferred over complex error recovery logic when the condition is extremely unlikely and the testing overhead to verify the recovery logic is high.

Only `RELEASE_ASSERT(condition)` should be used to validate conditions that might be imposed by the external environment. `ASSERT(condition)` should be used to document (and check in debug-only builds) program invariants.